Yosemite/Yosemite/Stores/ProductReviewStore.swift

```diff
@@ -0,0 +1,141 @@
+            retrieveProductReview(siteID: siteID, reviewID: reviewID, onCompletion: onCompletion)
+        }
+    }
+}
+
+
+// MARK: - Services!
+//
+private extension ProductReviewStore {
+
+    /// Deletes all of the Stored ProductReviews.
+    ///
+    func resetStoredProductReviews(onCompletion: () -> Void) {
+        let storage = storageManager.viewStorage
+        storage.deleteAllObjects(ofType: Storage.ProductReview.self)
+        storage.saveIfNeeded()
+        DDLogDebug("Product Reviews deleted")
+
+        onCompletion()
+    }
+
+    /// Synchronizes the product reviews associated with a given Site ID (if any!).
+    ///
+    func synchronizeProductReviews(siteID: Int, pageNumber: Int, pageSize: Int, onCompletion: @escaping (Error?) -> Void) {
+        let remote = ProductReviewsRemote(network: network)
+
+        remote.loadAllProductReviews(for: siteID, pageNumber: pageNumber, pageSize: pageSize) { [weak self] (productReviews, error) in
+            guard let productReviews = productReviews else {
+                onCompletion(error)
+                return
+            }
+
+            self?.upsertStoredProductReviewsInBackground(readOnlyProductReviews: productReviews) {
+                onCompletion(nil)
+            }
+        }
+    }
+
+    /// Retrieves the product review associated with a given siteID + reviewID (if any!).
+    ///
+    func retrieveProductReview(siteID: Int, reviewID: Int, onCompletion: @escaping (Networking.ProductReview?, Error?) -> Void) {
+        let remote = ProductReviewsRemote(network: network)
+
+        remote.loadProductReview(for: siteID, reviewID: reviewID) { [weak self] (productReview, error) in
+            guard let productReview = productReview else {
+                if case NetworkError.notFound? = error {
+                    self?.deleteStoredProductReview(siteID: siteID, reviewID: reviewID)
+                }
+                onCompletion(nil, error)
+                return
+            }
+
+            self?.upsertStoredProductReviewsInBackground(readOnlyProductReviews: [productReview]) {
+                onCompletion(productReview, nil)
+            }
+        }
+    }
+}
+
+
+// MARK: - Storage: ProductReview
+//
+private extension ProductReviewStore {
+
+    /// Deletes any Storage.ProductReview with the specified `siteID` and `reviewID`
+    ///
+    func deleteStoredProductReview(siteID: Int, reviewID: Int) {
+        let storage = storageManager.viewStorage
+        guard let productReview = storage.loadProductReview(siteID: siteID, reviewID: reviewID) else {
+            return
+        }
+
+        storage.deleteObject(productReview)
+        storage.saveIfNeeded()
+    }
+
+    /// Updates (OR Inserts) the specified ReadOnly ProductReview Entities *in a background thread*. onCompletion will be called
+    /// on the main thread!
+    ///
+    func upsertStoredProductReviewsInBackground(readOnlyProductReviews: [Networking.ProductReview], onCompletion: @escaping () -> Void) {
```

Comment:
I was wondering when we upsert the data in a background thread / core data context? vs. in `StatsStoreV4` we upsert the data to the view storage

also, for data deletion, looks like we do that in the view storage (as in `deleteStoredProductReview`). is this the pattern we generally follow: upsert in a derived background storage, and deletion in the view storage?

Reply:
I think it is, and I am not sure why/when it was decided to be it that way, but the way I see it, upserting is something that we do very often, and that we do on potentially large data sets, so it sounds reasonable to do it in a background thread. Deleting, however, only happens when logging out / switching stores so it's not that critical?

Reply:
Wow, it's been a long time since I've though about it. @jleandroperez or @bummytime ☝️  

...can probably explain the reasoning behind it. Last I recall, either the API didn't report deleted items, or our stack wasn't ready to handle deleting items, hence the view storage solution.

Reply:
The initial *Store* (concrete) implementations did most of the things in the ViewContext. 

**Why:** because the approach has always been... **let's try to keep things simple**, and complicate things up only when needed. Relatively quickly along the way, we've figured, several Upsert OPs should happen on a derived context.

As per deletions, there was no official rule. But i would always go with the same rule of thumb: is it an expensive OP? Background, otherwise... okay to do in the viewContext.

Deletions themselves usually are non-expensive. But if the entire transaction involves, maybe, nuking an entire dataset... perhaps a derivedContext is in place.

Creating a Derived Context is really fast, but when you effectively save the changes performed in BG... CoreData must push that into the ViewContext. 

Imagine a quick OP that takes milliseconds (think of ... maybe updating a single field). **Sometimes** getting things done right there is _faster_ than spawning another thread, and merging the changes back into the ViewContext. That's precisely what i've always tried to weight, all over!